Introduce general error handling (and other stuff)

bissim opened the pull request:
Fixes #6

As stated in #6, now every mission-critical command in `make_cluster.sh` script gets checked and if it fails execution is halted; it is a general case of the fix provided in PR #5 for issue #4.

I also reviewed English sentences along the scripts and README, [...]; moreover, I took the chance to make README  a bit more clearer, like removing a point in [_Dependencies_](https://github.com/isislab-unisa/aws-build-cluster-script/blob/master/README.md#dependencies) paragraph that is actually an anticipation of what's required in next paragraph and not an actual operative instruction (that would've led users to confusion).

Minor review in the sense to fix English has been apported to `state_cluster.sh` script as well.

`VERSION` has been bumped to 1.2.2 to eventually trigger any user of 1.2.0 and 1.2.1 to download the new version.

bissim committed:
Update make_cluster.sh
Fixes #6

- Reduce SSH connection attempts from 30 to 3
- Handle 'splash screen' so that it gets printed even in parameter mode
- Introduce utility function to handle generic errors and phase delimitation; in particular, 'checkError' function tells user which script phase failed, 'announcePhase' and 'announcePhaseTermination' print meaningful message to users to let him understand the execution progress
- Fix some ENGRISH both in code comments and in messages to user
- Use 'checkError' function after every 'ssh' or 'aws' command
- Integrate snippet from README to generate hostfiles and promt user for their creation

bissim committed:
Update state_cluster.sh
- Fix some ENGRISH
- Clearly state that EC2 instances are being handled with

bissim committed:
Update README.md
- Update formatting
- Fix ENGRISH
- Make mention in 'Dependencies' paragraph, point 1. of AWS CLI v2
- In point 2. of 'Dependencies' paragraph, inform user about configuration-related files for AWS CLI
- In 'Configure the script' (renamed from 'Config the script') point 4. could have been misleading, so it has been removed
- In 'Slave Nodes IPs Management' the range in the example has been fixed
- Get a hold of @labgua fixation over useless newlines all over the place
- Add a mention of myself (@bissim) as minor contributor

tl;dr MAKE README ACTUALLY READABLE

bissim committed:
Update README.md
- Actually rename 'Config the script' paragraph in 'Configure the script', as stated in f3031d7

bissim committed:
Update AWS Build Cluster Script to 1.2.2
- 0aea565: Update make_cluster.sh
- faeab1f: Update state_cluster.sh
- f3031d7: Update README.md
- c97f7c2: Update README.md

bissim commented:
A subtlety: in f3031d7 I also put version information in splash screen.

bissim commented on README.md:
```diff
@@ -1,19 +1,17 @@
-   Test if it works with the following command:
+1. Install the **AWS CLI**, see [official documentation](https://docs.aws.amazon.com/cli/latest/userguide/installing.html)
+   For AWS CLI v1, if you don't have it, you must install Python; for v2 all required tools are self-contained.
+   Test if it works by launching the following command:
```
Here, an user may be misled if it's using AWS CLI v2 that integrates Python and is therefore not needed in the environment. Quoting from [official source](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html):
> For AWS CLI version 2, it doesn't matter if you have Python installed and if you do, it doesn't matter which version. AWS CLI version 2 uses only the embedded version of Python (and any other dependencies) that is included in the installer.

bissim commented on README.md:
```diff
@@ -79,27 +79,21 @@ To use this script you must have AWS CLI and jq
-   Now you can choose what kind of machine: you must take note of the code name.
-   See the [catalog](https://aws.amazon.com/ec2/instance-types/)
-
-
```
This whole point is actually a requirement for _Create a cluster_ paragraph: that's the reason the catalog link has been moved to line 88.

bissim commented on README.md:
```diff
@@ -237,23 +229,25 @@ This command should show a json response with the state of all cluster instances
-| ami-52a0c53b |           starcluster-base-ubuntu-12.04-x86_64-hvm          |  Ubuntu |    T2.Micro   |  WORK  |
-| ami-116d857a |         debian-jessie-amd64-hvm-2015-06-07-12-27-ebs        |  Debian |    T2.Micro   |  WORK  |
-| ami-00035c7b | Fedora-Atomic-25-20170727.0.x86_64-us-east-1-HVM-standard-0 |  Fedora |    T2.Micro   |  WORK  |
+|:------------:|:-----------------------------------------------------------:|:-------:|:-------------:|:-------:|
```
Whoops, I forgot two dashes here under `INSTANCE TYPE`.

bissim committed:
Update make_cluster.sh
- Properly empty hostfiles

labgua commented:
> Fixes #6
> 
> As stated in #6, now every mission-critical command in `make_cluster.sh` script gets checked and if it fails execution is halted; it is a general case of the fix provided in PR #5 for issue #4.
> 
> I also reviewed English sentences along the scripts and README, since @labgua is not very fond with English (eh Sergio?); moreover, I took the chance to make README a bit more clearer, like removing a point in [_Dependencies_](https://github.com/isislab-unisa/aws-build-cluster-script/blob/master/README.md#dependencies) paragraph that is actually an anticipation of what's required in next paragraph and not an actual operative instruction (that would've led users to confusion).
> 
> Minor review in the sense to fix English has been apported to `state_cluster.sh` script as well.
> 
> `VERSION` has been bumped to 1.2.2 to eventually trigger any user of 1.2.0 and 1.2.1 to download the new version.

Salve @bissim 
La ringrazio per il suo contributo ma nessuno le chiede di aiutarci e nessuno le da il diritto di giudicare: se le và posti pure la tua soluzione.
Se non capisce cosa c'è scritto in questo commento le do un consiglio: usi google traduttore... sempre ammesso che sia questo il suo problema.

Here the link for the translation https://tinyurl.com/translbissim

bissim commented:
After a brief review of my message (and a little bit of search in my dictionary), I came up to the source of this misunderstanding in my use of the expression _to be fond of_ that has an actual stronger meaning than the one I intended (still, it strangely gets translated nicely by Google Translator).
The meaning I intended was 'not to feel at one's ease'.

That being said, not feeling at one's ease with English is not something one should be ashamed of or feel guilt for, nor this is something I can feel responsible for in any degree or in any other way.
This is my last stance over this matter.

So, it's up to you to decide whether merge or reject this PR; I'm fine with whatever you decide.

bissim committed:
Update make_cluster.sh
- Add generation of public IPs array file
- Remove previously inserted snippet to generate hostfiles, not actually needed in this context

bissim committed:
Speed up cluster generation
Significantly reduce execution time by setting most of remote commands in for loops as background processes but still monitoring their outcome

bissim committed:
Update AWS Build Cluster Script to 1.3.0
All the effort and results in 5691ddf deserve a minor version increase and some more acknowledgement

bissim committed:
Update make_cluster.sh
Revert /etc/hosts file manipulation

bissim committed:
Update AWS Build Cluster Script to 1.3.1

bissim committed:
Update make_cluster.sh
Make SSH test connection parallel as well

labgua commented:
Thanks a lot for your contribution.
I'll return the favour.
